include/output.hxx

```diff
@@ -50,64 +49,211 @@ using std::endl;
+
+  virtual void enable();  ///< Enables writing to stdout (default)
+  virtual void disable(); ///< Disables stdout
 
   int open(const char *fname, ...); ///< Open an output log file
-  void close();                    ///< Close the log file
+  void close();                     ///< Close the log file
+
+  virtual void write(const char *string, ...); ///< Write a string using C printf format
+
+  virtual void print(const char *string, ...); ///< Same as write, but only to screen
 
-  void write(const char*string, ...); ///< Write a string using C printf format
+  virtual void vwrite(const char *string,
+                      va_list args); ///< Write a string using C vprintf format
 
-  void print(const char*string, ...); ///< Same as write, but only to screen
+  virtual void vprint(const char *string,
+                      va_list args); ///< Same as vwrite, but only to screen
 
   /// Add an output stream. All output will be sent to all streams
-  void add(std::basic_ostream<char, _Tr>& str) {
-    multioutbuf_init::buf()->add(str);
-  }
+  void add(std::basic_ostream<char, _Tr> &str) { multioutbuf_init::buf()->add(str); }
 
   /// Remove an output stream
-  void remove(std::basic_ostream<char, _Tr>& str) {
+  void remove(std::basic_ostream<char, _Tr> &str) {
     multioutbuf_init::buf()->remove(str);
   }
 
   static Output *getInstance(); ///< Return pointer to instance
-  static void cleanup();   ///< Delete the instance
- private:
+  static void cleanup();        ///< Delete the instance
+
+private:
   static Output *instance; ///< Default instance of this class
+
+  std::ofstream file;                 ///< Log file stream
+  static const int BUFFER_LEN = 1024; ///< default length
+  int buffer_len;                     ///< the current length
+  char *buffer;                       ///< Buffer used for C style output
+  bool enabled;                       ///< Whether output to stdout is enabled
+};
+
+/// Class which behaves like Output, but has no effect.
+/// This is to allow debug outputs to be disabled at compile time
+/// 
+/// 
+class DummyOutput : public Output {
+public:
+  void write(const char *str, ...) override {};
+  void print(const char *str, ...) override {};
+  void enable() override {
+    throw BoutException("DummyOutput cannot be enabled.\nTry compiling with "
+                        "--enable-debug or be less verbose?");
+  };
+  void disable() override{};
+  void enable(bool en) {
+    if (en)
+      this->enable();
+  };
+};
+
+/// Layer on top of Output which passes through calls to write, print etc
+/// if it is enabled, but discards messages otherwise.
+/// This is used to provide different levels of output
+/// (info, prog, warn, error) which can be enabled and disabled at run time.
+///
+class ConditionalOutput : public Output {
+public:
+  /// @params[in] base    The Output object which will be written to if enabled
+  ConditionalOutput(Output *base_) : base(base_), enabled(true), base_is_cond(false) {};
```

Comment:
`base_` could just be `base` -- it's not used in the constructor body, so no chance for confusion. It would also make it match the doxygen comment

Reply:
Done